lab/lab-guide.md

```diff
@@ -0,0 +1,1032 @@
+    string queryText = $"""
+        SELECT Top @maxContextWindow
+            *
+        FROM c  
+        WHERE 
+            c.tenantId = @tenantId AND 
+            c.userId = @userId AND
+            c.sessionId = @sessionId AND 
+            c.type = @type
+        ORDER BY 
+            c.timeStamp DESC
+        """;
+    ```
+
+    After querying for the most recent messages in Azure Cosmos DB, we reverse put them back in order in reverse. Recency matters in a conversation, the most recent text is what we want closer to the actual question. Counting the number of messages allows us to control the total number of tokens used while still providing relevant context.
+
+1. Next, within the **ChatService.cs** class, locate **GetChatCompletionAsync()** with the following signature. 
+
+    ```csharp
+    public async Task<Message> GetChatCompletionAsync(string tenantId, string userId, string sessionId, string promptText)
+    ```
+
+1. Comment out the two lines below initializing the **messages** variable and passing it to the Semantic Kernel service. Then add the two new lines as seen below. This calls the the function to get the context window that we just updated, and passes the context window to the Semantic Kernel Service to get a completion from Azure OpenAI.
+
+    ```csharp
+    //List<Message> messages = new List<Message>() { chatMessage };
+    //(chatMessage.Completion, chatMessage.CompletionTokens) = await _semanticKernelService.GetChatCompletionAsync(messages);
+
+    //Get the context window for this conversation up to the maximum conversation depth.
+    List<Message> contextWindow = 
+        await _cosmosDbService.GetSessionContextWindowAsync(tenantId, userId, sessionId, _maxContextWindow);
+    (chatMessage.Completion, chatMessage.CompletionTokens) = await _semanticKernelService.GetChatCompletionAsync(contextWindow);
+    ```
+
+    Now we are passing in a list of messages that represents the conversation history in order to get our contextually relevant completion. This is why we had to implement **GetChatCompletionAsync()** to take a list of prompts, rather than taking a single prompt representing the current user message alone.
+
+1. Save the **Services/ChatService.cs** file.
+
+## Check your work
+
+You are now ready to test your context window implementation.
+
+1. Back in our terminal, start the application using **dotnet run**.
+
+    ```bash
+    dotnet run
+    ```
+
+1. Use `CTRL + click` to open the .NET Aspire dashboard. Once the dashboard loads, click the link to open our chat application.
+
+1. In the web application, create a new chat session and ask the AI assistant this question, `What is the most expensive bike?`. The AI assistant now responds with a completion created by the model saying that "24K Gold Extreme Mountain Bike" is the most expensive, with some additional information. Now the second question again, `What about the least expensive?`. You should see some inexpensive bike options.
```

Comment:
@jcocchi how does the LLM know to respond this way? The system prompt isn't very specific here.

Reply:
This is the answer I got after a few rounds of testing, do you get different results? We can certainly tweak all the prompts for this lab, I was hoping to get feedback on the question flow